### PR TITLE
PC: Housekeeping

### PR DESCRIPTION
**Pull request by pbrubeck**
# Description

`PCSNESBase` implements `form()` to remove code duplication.

`FacetSplitPC` now supports a split onto any number of restricted elements (this allows proper subspaces).

`HiptmairPC` now supports restricted potential spaces.

`tabulate_exterior_derivative` works for simplices.

### CODE REVIEW

File: firedrake/preconditioners/assembled.py [hunk 1]
```diff
@@ -29,15 +26,11 @@ def initialize(self, pc):
...
-        if len(V) == 1:
-            V = FunctionSpace(V.mesh(), V.ufl_element())
-        else:
-            V = MixedFunctionSpace([V_ for V_ in V])
+        V = get_function_space(pc.getDM()).collapse()
         test = TestFunction(V)
         trial = TrialFunction(V)
 
-        if P.type == "python":
+        if P.getType() == "python":
```
Comment (dham) [inline firedrake/preconditioners/assembled.py:33]: please stop programming in Java.

File: firedrake/preconditioners/base.py [hunk 2]
```diff
@@ -58,9 +59,44 @@ def destroy(self, pc):
...
+        application-context, if desired.
+
+        :arg obj: a `PETSc.PC` or `PETSc.SNES` object.
+
+        :arg test: a `TestFunction` on this `FunctionSpace`.
+
+        :arg trial: a `TrialFunction` on this `FunctionSpace`.
+
+        :returns `(a, bcs)`, where `a` is a bilinear `Form`
+        and `bcs` is a list of `DirichletBC` boundary conditions (possibly `None`).
```
Comment (dham) [inline firedrake/preconditioners/base.py:76]: Numpydoc style docstring.

File: firedrake/preconditioners/facet_split.py [hunk 3]
```diff
@@ -25,94 +26,75 @@ class FacetSplitPC(PCBase):
...
-        assert len(V) == 1, "Interior-facet decomposition of mixed elements is not supported"
-
-        def restrict(ele, restriction_domain):
-            if isinstance(ele, VectorElement):
-                return type(ele)(restrict(ele._sub_element, restriction_domain), dim=ele.num_sub_elements)
-            elif isinstance(ele, TensorElement):
-                return type(ele)(restrict(ele._sub_element, restriction_domain), shape=ele._shape, symmetry=ele._symmety)
-            else:
-                return RestrictedElement(ele, restriction_domain)
+        assert len(V) == 1, "Decomposition of mixed elements is not supported"
```
Comment (dham) [inline firedrake/preconditioners/facet_split.py:55]: please to be exception.

File: firedrake/preconditioners/facet_split.py [hunk 4]
```diff
@@ -25,94 +26,75 @@ class FacetSplitPC(PCBase):
...
-            mixed_opmat.setTransposeNullSpace(_permute_nullspace(P.getTransposeNullSpace()))
-        elif self.perm:
-            global_indices = V.dof_dset.lgmap.apply(self.iperm.indices)
+            self._assemble_mixed_op(tensor=self.P)
+            self.mixed_opmat = self.P.petscmat
+            self.set_nullspaces(pc)
+            self.work_vecs = self.mixed_opmat.createVecs()
+        elif self.subset:
+            global_indices = V.dof_dset.lgmap.apply(self.subset.indices)
             self._global_iperm = PETSc.IS().createGeneral(global_indices, comm=P.getComm())
```
Comment (dham) [inline firedrake/preconditioners/facet_split.py:91]: use `self.comm`?

**Review by dham — CHANGES_REQUESTED** [review]

File: firedrake/preconditioners/base.py (hunk 2 above)
Comment (JDBetteridge) [reply firedrake/preconditioners/base.py:76]: https://github.com/firedrakeproject/firedrake/wiki/Docstrings

**Review by dham — APPROVED** [review]

**Pull request merged**